### ISSUE TITLE
hardhat compile complains about contract size to big even if it's not

### ISSUE THREAD

**Issue by migoldfinger** [NomicFoundation/hardhat#4025]
### Version of Hardhat

2.15.0

### What happened?

hardhat compile states
```
Solidity 0.8.20 is not fully supported yet. You can still use Hardhat, but some features, like stack traces, might not work correctly.

Learn more at https://hardhat.org/hardhat-runner/docs/reference/solidity-support

Warning: Contract code size is 27806 bytes and exceeds 24576 bytes (a limit introduced in Spurious Dragon). This contract may not be deployable on Mainnet. Consider enabling the optimizer (with a low "runs" value!), turning off revert strings, or using libraries.
   --> contracts/core/Vault.sol:144:1:
    |
144 | contract Vault is ReentrancyGuard, IVault {
    | ^ (Relevant source part starts here and spans across multiple lines).


Generating typings for: 6 artifacts in dir: typechain-types for target: ethers-v5
Successfully generated 370 typings!
Compiled 3 Solidity files successfully
```
I added hardhat-contract-sizer and the output for the contract in question is
```
 |  Vault                 ·                   22.363  ·         0.000  │
```
which is perfectly fine and under the 24576 bytes limit. And deployment is working at least for the testnets

I suspect there is a place where config override is ignored since my config defines a override for this specific contract
```
solidity: {
		compilers: [
			{
				version: "0.8.20",
				settings:
				{
					optimizer:
					{
						enabled: true,
						runs: 500000
					}
				}
			},
			{
				version: "0.6.12",
				settings:
				{
					optimizer:
					{
						enabled: true,
						runs: 10
					}
				}
			}
		],
		overrides: {
			"contracts/core/Vault.sol": {
				version: "0.8.20",
				settings:
				{
					optimizer:
					{
						enabled: true,
						runs: 200
					}
				}
			}
		}
```
I came to this conclusion since since the values are identical if I remove the override setting or change the runs value to 500000

### Minimal reproduction steps

1) Create a contract with a high runs value that exceeds the contract size limit
2) Make sure the contract size is lower than the size limit with a lower runs value
3) create a configuration with the high runs value as default and the lower runs setting as override
4) run hardhat compile
5) optional check contract sizes with hardhat-contract-sizer plugin

### Search terms

compile size limit warning

**Comment by fvictorio** [comment]
First, a clarification: that warning comes from solc, not from hardhat.

I'm not familiar with the internals of `hardhat-contract-sizer`, but I wrote a big contract, and got this warning:

```
Warning: Contract code size is 32029 bytes and exceeds 24576 bytes (a limit introduced in Spurious Dragon). This contract may not be deployable on Mainnet. Consider enabling the optimizer (with a low "runs" value!), turning off revert strings, or using libraries.
```

Running `npx hardhat size-contracts` produces this:

```
 ·------------------------|--------------------------------|--------------------------------·
 |  Solc version: 0.8.17  ·  Optimizer enabled: false      ·  Runs: 200                     │
 ·························|································|·································
 |  Contract Name         ·  Deployed size (KiB) (change)  ·  Initcode size (KiB) (change)  │
 ·························|································|·································
 |  Foo                   ·                31.278 (0.000)  ·                33.189 (0.000)  │
 ·------------------------|--------------------------------|--------------------------------·
```

But according to the generated artifacts, solc is right:

```
❯ jq '(.deployedBytecode[2:] | length) / 2' artifacts/contracts/Foo.sol/Foo.json
32029
```

So... I don't know? @ItsNickBarry do you know what could be going on here?

**Comment by ItsNickBarry** [comment]
By default, `hardhat-contract-sizer` outputs values in KiB, because the limit is exactly 24 KiB.  32029 bytes is equal to 31.278 kibibytes, so the output posted by @fvictorio is correct.  I have no idea where the discrepancy posted by @migoldfinger might have come from.

**Comment by fvictorio** [comment]
I see, thanks for chiming in! In that case, I think I'm going to close this issue for the time being. I'll re-open if there is some underlying problem in Hardhat that is causing this.

**Comment by migoldfinger** [comment]
You are all missing the point...
I think the discrepancy is when using overrides in the config.

My guess is hardhat is compiling all contracts with the default settings first. solc then will show this warning.
Then it will compile the override files with the override settings.
Using the **overrides** section is crucial for this.

To be clear in the end all works as expected but the warning is misleading and wrong.
The warning would be true if there was no **overrides** section in the config.
But since there is and the settings there ensures that contract size goes under the limit the warning is plain wrong.
The solution would be that files addressed in the **overrides** section should not get compiled with the default settings from default section.
You can use https://github.com/DFX-Protocol/dfx-contracts for quick reproduction

@fvictorio Since this is clearly not closed please reopen.

**Comment by fvictorio** [comment]
Ok, this is a minimal reproduction of this problem.

First create two contracts: `Foo.sol` and `Bar.sol`:

```solidity
// Foo.sol

import "./Bar.sol";

contract Foo {
  Bar public bar;
}
```

```solidity
// Bar.sol

contract Bar {
  string s = "";

  function f1() public {
    s = "asdf";
  }

  function f2() public {
    s = "asdf";
  }

  // ...297 functions more like this...

  function f300() public {
    s = "asdf";
  }
}
```

There are three compilation scenarios here (I used solc v0.8.17 here):

1. If you compile the project with the optimizer disabled, you'll get the contract code size warning. This is correct.
2. If you compile the project with the optimizer enabled, you won't get the warning. This is also correct.
3. If you compile the project with the optimizer disabled BUT use an override to enable the optimizer for `Bar.sol`, you'll get the warning.

The reason scenario 3 happens is because there are two compilation jobs being executed here:

- First, both `Foo.sol` and `Bar.sol` are compiled together, with the optimizer disabled. This generated the warning, but the result of this compilation is only used to generate the bytecode for `Foo`, so the warning is misleading.
- Then, `Bar.sol` is compiled on its own, with the optimizer enabled, and the resulting bytecode is used for `Bar`

The fix here is something that we actually considered doing once: only include the relevant contracts in the `outputSelection` entry of the compilation input. We didn't end up doing it because the performance gains didn't seem worth it, but this issue might be a good reason to do it anyway.

**Comment by migoldfinger** [comment]
Yes, that is exactly what happens. Thanks for the detailed explanation.
